### PR TITLE
Add TypeScript UserService and AuthService implementations using MongoDB

### PR DESCRIPTION
**Pull request by sherryhli**
## Notion ticket link

* Addresses the user service part of [Firebase authentication & user services using MongoDB (Mongoose)](https://www.notion.so/uwblueprintexecs/Firebase-authentication-user-services-using-MongoDB-Mongoose-1618340d0baa4549a8e38cd521c149c0)
* [Configure TS logger for backend](https://www.notion.so/uwblueprintexecs/Configure-TS-logger-for-backend-f4234bb4329143548896825310127acb)

## Implementation description
* Implemented methods defined in `IUserService`
* Defined MongoDB `User` model using Mongoose
* Integrated Firebase Admin SDK for auth related user management logic
* Set up logger using [winston](https://www.npmjs.com/package/winston), which is useful for keeping track of possible inconsistencies across Firebase and MongoDB user data (caused by rollback failure edge cases)

## Steps to test
1. Generate a Firebase service account private key
    * Go to our project in the [Firebase console](https://console.firebase.google.com)
    * "Project settings" > "Service accounts" > "Generate private key"
    * A file will be downloaded, copy it into `/backend/typescript/` and rename it **`firebaseServiceAccount.json`**
 2. Pull from secret manager **OR** add this line to the root `.env` file in the repo
 ```
 GOOGLE_APPLICATION_CREDENTIALS=firebaseServiceAccount.json
 ```
 3. Comment out the `frontend` and `py-backend` services in `docker-compose.yml`
 4. Run the application:
 ```
docker-compose up --build
```
5. Send requests to the test endpoints in `backend/typescript/server.ts` (note they are all GET requests). Suggested test plan:
    * Create some users by sending requests to `/create-user/:firstName/:lastName`.
    * Update some users sending requests to `/update-user/:userId`
    * Retrieve a user by `userId` by sending a request to `/get-user-by-id/:userId`
    * Retrieve a user by `email` by sending a request to `/get-user-by-id/:email`
    * Retrieve all users by sending a request to `/get-users`
    * Delete a user by `userId` by sending a request to `/delete-user-by-id/:userId`
    * Delete a user by `email` by sending a request to `/delete-user-by-email/:email`

Please check the data in MongoDB Atlas and the Firebase console as well.

## What should reviewers focus on?
* The most important file to look at is `backend/typescript/services/implementations/userService.ts`
* Error handling and "rollbacks": does the logic make sense, are there cleaner ways to do this?
* Concerns on data consistency - seems like an unavoidable problem when we use both Firebase and another data store, but let me know if you think of a better way!
* Am I using TypeScript features properly?

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s) (note: unused var error on `IAuthService` is expected)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/typescript/models/person.mgmodel.ts [hunk 1]
```diff
@@ -1,19 +0,0 @@
-import mongoose, { Schema, Document } from "mongoose";
```
Comment (sherryhli) [inline backend/typescript/models/person.mgmodel.ts:1]: Removed this file since we have a real MongoDB model now

File: backend/typescript/server.ts [hunk 2]
```diff
@@ -1,24 +1,71 @@
...
+import * as firebaseAdmin from "firebase-admin";
 
-import { mongo, sequelize } from "./models/index";
-// import PgPerson from "./models/person.pgmodel";
-// import MgPerson from "./models/person.mgmodel";
+import { mongo, sequelize } from "./models";
+
+/** ***************************************************************************
+ * TEMPORARY: UserService test endpoints
+ *************************************************************************** */
```
Comment (sherryhli) [inline backend/typescript/server.ts:9]: Just some test code, feel free to skip this part when reviewing

File: backend/typescript/services/implementations/userService.ts [hunk 3]
```diff
@@ -0,0 +1,240 @@
+import * as firebaseAdmin from "firebase-admin";
+
+import IUserService from "../interfaces/userService";
+import MgUser, { User } from "../../models/user.mgmodel";
+import { CreateUserDTO, UpdateUserDTO, UserDTO } from "../../types";
+import Logger from "../../utilities/logger";
+
+class UserService implements IUserService {
+  /* eslint-disable class-methods-use-this */
```
Comment (sherryhli) [inline backend/typescript/services/implementations/userService.ts:9]: We're not using any class properties so all the methods here can be made `static`, but I want to stick with our design of using the interface.

File: backend/typescript/services/interfaces/userService.ts [hunk 4]
```diff
@@ -5,47 +5,49 @@ interface IUserService {
...
    * @param userId user's userId
+   * @throws Error if user deletion fails
    */
-  deleteUserById(userId: number): void;
-
-  /**
-   * Delete a user by email
-   * @param email user's email
-   */
-  deleteUserByEmail(email: string): void;
```
Comment (sherryhli) [inline backend/typescript/services/interfaces/userService.ts:50]: Deleted this since I originally thought there was a blocker for doing rollbacks, but on second thought I think it's a non-issue, will add it back.

File: backend/typescript/types.ts [hunk 5]
```diff
@@ -1,12 +1,12 @@
-export type Role = number;
+export type Role = "User" | "Admin";
```
Comment (sherryhli) [inline backend/typescript/types.ts:1]: This makes things easy with MongoDB since enums are stored as strings, but open to making this an actual Enum if that makes more sense for the Postgres implementation @zm2he

File: backend/typescript/types.ts [hunk 6]
```diff
@@ -1,12 +1,12 @@
...
 
 export type Token = {
   accessToken: string;
   refreshToken: string;
 };
 
 export type UserDTO = {
-  id: number;
+  id: string;
```
Comment (sherryhli) [inline backend/typescript/types.ts:9]: Using the string representaiton of the MongoDB object ID. Postgres implementation will still use a number @zm2he

File: backend/typescript/utilities/logger.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+if (process.env.NODE_ENV !== "production") {
+  logger.add(
+    new winston.transports.Console({
+      format: winston.format.combine(
+        winston.format.timestamp(),
+        winston.format.json(),
+      ),
+    }),
+  );
+}
```
Comment (sherryhli) [inline backend/typescript/utilities/logger.ts:24]: Log into the files `error.log` and `combined.log`, and also to the console when not in prod.

Format of log message:
```
{"message":"......","level":"error","timestamp":"2021-03-13T03:26:25.269Z"}
```

File: backend/typescript/services/implementations/userService.ts [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+    let firebaseUser: firebaseAdmin.auth.UserRecord;
+
+    try {
+      user = await MgUser.findById(userId);
+
+      if (!user) {
+        throw new Error(`userId ${userId} not found.`);
+      }
+
+      firebaseUser = await firebaseAdmin.auth().getUser(user.authId);
```
Comment (sherryhli) [inline backend/typescript/services/implementations/userService.ts:21]: The Firebase query is just to get the email, we could store email in MongoDB as well to avoid doing this for all the gets, but decided not to prematurely optimize (also I think duplicating the data will further complicate consistency issues)

Comment (alexguo8) [reply backend/typescript/services/implementations/userService.ts:21]: This seems ok, as long as there are not a lot of cases where we need the email for non-auth purposes

**Review by xinhaoz — DISMISSED** [review]
LGTM ✨

**Review by raveen-singh — DISMISSED** [review]
LGTM!

**Review by carelynntsai — APPROVED** [review]
